pkg/informers/informers.go

```diff
@@ -0,0 +1,180 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+)
+
+type GenericInformer interface {
+	Informer() cache.SharedIndexInformer
+	Lister() cache.GenericLister
+}
+
+// InformerFactoriesForNamespaces is a simple way to combine several shared informers into a single struct with unified listing power
+type InformerFactoriesForNamespaces interface {
+	ForResource(namespace string, resource schema.GroupVersionResource) (GenericInformer, error)
+	Namespaces() sets.String
+}
+
+type InformersForResource struct {
+	informers []GenericInformer
+}
+
+func NewInformersForResource(ifs InformerFactoriesForNamespaces, resource schema.GroupVersionResource) (*InformersForResource, error) {
+	namespaces := ifs.Namespaces().List()
+	var informers []GenericInformer
+
+	for _, ns := range namespaces {
+		informer, err := ifs.ForResource(ns, resource)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error getting informer for resource %v", resource)
+		}
+		informers = append(informers, informer)
+	}
+
+	return &InformersForResource{
+		informers: informers,
+	}, nil
+}
+
+func (w *InformersForResource) Start(stopCh <-chan struct{}) {
+	for _, i := range w.informers {
+		go i.Informer().Run(stopCh)
+	}
+}
+
+func (w *InformersForResource) GetInformers() []GenericInformer {
+	return w.informers
+}
+
+func (w *InformersForResource) AddEventHandler(handler cache.ResourceEventHandler) {
+	for _, i := range w.informers {
+		i.Informer().AddEventHandler(handler)
+	}
+}
+
+func (w *InformersForResource) AddEventHandlerWithResyncPeriod(handler cache.ResourceEventHandler, resyncPeriod time.Duration) {
+	for _, i := range w.informers {
+		i.Informer().AddEventHandlerWithResyncPeriod(handler, resyncPeriod)
+	}
+}
+
+func (w *InformersForResource) HasSynced() bool {
+	for _, i := range w.informers {
+		if !i.Informer().HasSynced() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (w *InformersForResource) List(selector labels.Selector) ([]runtime.Object, error) {
+	var ret []runtime.Object
+
+	for _, inf := range w.informers {
+		objs, err := inf.Lister().List(selector)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, objs...)
+	}
+
+	return ret, nil
+}
+
+func (w *InformersForResource) ListAllByNamespace(namespace string, selector labels.Selector, appendFn cache.AppendFunc) error {
+	for _, inf := range w.informers {
+		err := cache.ListAllByNamespace(inf.Informer().GetIndexer(), namespace, selector, appendFn)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (w *InformersForResource) Get(name string) (runtime.Object, error) {
+	var err error
+
+	for _, inf := range w.informers {
+		var ret runtime.Object
+		ret, err = inf.Lister().Get(name)
+
+		if apierrors.IsNotFound(err) {
```

Comment:
I'm not sure this is correct. What if a name doesn't exist in ANY informer? Then we will continue and eventually return an error that says that the error was not found in one _particular_ namespace, i.e. the last informer that was checked, which may not be the namespace the caller actually cares about.

Reply:
`Get` is invoked with `[ns]/[name]`. client-go does not inject the namespace of the underlying informer but keeps the requested namespace. I coded a quick example, where `ns1`, `ns2`, `ns3` all contain a servicemonitor called `prometheus-example-monitor`:

When we initialize the informers with `ns1` and `ns2` only:
```
	infs, err := informers.NewInformersForResource(
		informers.NewMonitoringInformerFactories(
			map[string]struct{}{
				"ns1": {},
				"ns2": {},
			}, nil,
			mclient, 5*time.Minute, nil,
		),
		monitoringv1.SchemeGroupVersion.WithResource(monitoringv1.ServiceMonitorName),
	)
```

And look for `ns3/prometheus-example-monitor` 
```
	ret, err := infs.Get("ns3/prometheus-example-monitor")
```
the following error is being returned by client-go:
```
    informers_test.go:63: servicemonitors.monitoring.coreos.com "ns3/prometheus-example-monitor" not found
```

You are right that the error is returned by the `"ns2"` informer, but the error message is the same as for the `"ns1"` informer.

Reply:
However, we could construct our own explicit apierror to avoid confusion, wdyt?

Reply:
Ack; let's keep this in mind but this is not blocking. At the heart of this, because we are doing a lot of magic, we may need to build our own errors too in order to avoid a leaky abstraction. E.g. say that next week client-go changes the error message to be "resource x/y could not be found in namespace z"; this would be confusing to users asking about resource "y" in namespace "x".